graphql/graphql.go

```diff
@@ -281,40 +281,40 @@ func (t *Transaction) EffectiveGasPrice(ctx context.Context) (*hexutil.Big, erro
 	return (*hexutil.Big)(math.BigMin(new(big.Int).Add(tx.GasTipCap(), header.BaseFee), tx.GasFeeCap())), nil
 }
 
-func (t *Transaction) MaxFeePerGas(ctx context.Context) (*hexutil.Big, error) {
-	tx, _, err := t.resolve(ctx)
-	if err != nil || tx == nil {
-		return nil, err
+func (t *Transaction) MaxFeePerGas(ctx context.Context) *hexutil.Big {
+	tx, _ := t.resolve(ctx)
```

Comment:
This seems wrong. The `resolve` on a transaction ought to be able to return an error, I think. The fact that it currently does not looks like a flaw to me. 
```
	tx, blockHash, _, index, err := t.r.backend.GetTransaction(ctx, t.hash)
	if err == nil && tx != nil {
             ...
	}
	// No finalized transaction, try to retrieve it from the pool
	t.tx = t.r.backend.GetPoolTransaction(t.hash)
	return t.tx, nil, nil
}
```
We just "forget" the `err`

Reply:
make sense, Try to add an error handling that can't find tx in txpool now @holiman

Reply:
> This seems wrong. The resolve on a transaction ought to be able to return an error, I think. The fact that it currently does not looks like a flaw to me.

The idea here is that we try to resolve and return nil to user if no such tx was found. This is similar behavior to `eth.getTransactionByHash`.

So IMO the initial PR that removed the error param is ok. Only we can add a comment to clarify this point. And make sure nowhere we can trigger a panic on nil.

Reply:
> > This seems wrong. The resolve on a transaction ought to be able to return an error, I think. The fact that it currently does not looks like a flaw to me.
> 
> The idea here is that we try to resolve and return nil to user if no such tx was found. This is similar behavior to `eth.getTransactionByHash`.
> 
> So IMO the initial PR that removed the error param is ok. Only we can add a comment to clarify this point. And make sure nowhere we can trigger a panic on nil.

Already revert commit. BTW  `getTransactionByHash()` also has this problem. `HeaderByhash()` and `GetTransaction()` in eth/api_backend.go also always return a nil as the return value of error. To be precise, there are many such problems in **api_backend**, such as 'getpoolTransactions()'. I will open a new pr to fix **api_backend.go**

Reply:
> The idea here is that we try to resolve and return nil to user if no such tx was found. This is similar behavior to eth.getTransactionByHash.

That might be fine for a lot of cases, such as this right here. But IMO the case where this fails are these: 

```
func (t *Transaction) Nonce(ctx context.Context) hexutil.Uint64 
```
So the `Nonce` will return `0` if the transaction cannot be found. It cannot return `nil`, to signal "could not find it", and if we don't have the ability to return an error, we'll just return a wrong answer. Might be only `Nonce`, `Gas` and `InputData` that are affected.

Reply:
I think I understand what you mean now. Even though we have addressed the issue of returning nil error, we should also pay attention to the problem of returning 0 when certain data like nonce is not found in a transaction (which can be misleading as if nonce is found to be 0). Instead, we should provide a prompt indicating the failure to find the requested data.

Reply:
> > The idea here is that we try to resolve and return nil to user if no such tx was found. This is similar behavior to eth.getTransactionByHash.
> 
> 
> 
> That might be fine for a lot of cases, such as this right here. But IMO the case where this fails are these: 
> 
> 
> 
> ```
> 
> func (t *Transaction) Nonce(ctx context.Context) hexutil.Uint64 
> 
> func (t *Transaction) InputData(ctx context.Context) hexutil.Bytes 
> 
> func (t *Transaction) Gas(ctx context.Context) hexutil.Uint64 
> 
> ```
> 
> So the `Nonce` will return `0` if the transaction cannot be found. It cannot return `nil`, to signal "could not find it", and if we don't have the ability to return an error, we'll just return a wrong answer. Might be only `Nonce`, `Gas` and `InputData` that are affected. 

@holiman l think I understand what you mean now. Even though we have addressed the issue of returning nil error, we should also pay attention to the problem of returning 0 when certain data like nonce is not found in a transaction (which can be misleading as if nonce is found to be 0). Instead, we should provide a prompt indicating the failure to find the requested data.

Reply:
Ah I thought I responded yesterday. That shouldn't be a problem. Because if a tx doesn't exist we will not have a Transaction object in the response anyway. Here are the various ways you could query a tx within the schema:

- Query directly by hash -> [returns nil](https://github.com/ethereum/go-ethereum/blob/ac86547b0154129f4dac865579946bd2e9378fb0/graphql/graphql.go#LL1241C1-L1247C3) if non-existent
- Get transactions of a block -> implies tx exists
- Get pending txes -> implies tx exists
- Given a Log, get transaction that emitted it -> implies tx exists

Reply:
> Ah I thought I responded yesterday. That shouldn't be a problem. Because if a tx doesn't exist we will not have a Transaction object in the response anyway. Here are the various ways you could query a tx within the schema:
> 
> * Query directly by hash -> [returns nil](https://github.com/ethereum/go-ethereum/blob/ac86547b0154129f4dac865579946bd2e9378fb0/graphql/graphql.go#LL1241C1-L1247C3) if non-existent
> * Get transactions of a block -> implies tx exists
> * Get pending txes -> implies tx exists
> * Given a Log, get transaction that emitted it -> implies tx exists

so this version is still cool ?  https://github.com/ethereum/go-ethereum/pull/27285/commits/c8f93f804252e4380b08488b9dbb18b033108976